### PR TITLE
Fall back to row-based evaluation of regular expressions using method derived from evaluating UDFs

### LINKED ISSUE #7001: [FEA] Fallback to row-based (CPU) evaluation of regexp expressions in context of GPU
**Is your feature request related to a problem? Please describe.**
From this epic (#6955):

* We currently have `spark.rapids.sql.rowBasedUDF.enabled` that lets us put a UDF on the CPU, but do so with the GPU semaphore held. We should look at what the performance impact is for something like this but for a more concrete operator, like regexp_extract. We know that there are a lot of regular expression cases that we cannot currently support so regular expressions is a good starting point for this.

**Describe the solution you'd like**
Fallback to GPU using row-based UDF approach for `regexp_*`-based expressions.

### PR DESCRIPTION
**Pull request by NVnavkumar**
Fixes #7001.

This adds `GpuWrappedRowBasedExpression` which allows for the row-based evaluation of Spark `Expression` classes that are usually applied to row. For now, these are used to wrap regExp Expressions from Spark (`RLike`, `RegExpExtract`, `RegExpExtractAll`, `RegExpReplace`, and `Split`), and are used when the plugin sees regular expressions that we cannot support on the GPU. This allows us to support effectively all regular expressions that Spark supports and still keep the underlying node on the GPU (holding the semaphore) while the expression can be evaluated on the CPU.

Some unit tests have been added for this support and integration tests have been updated to handle regular expressions with lookahead (a feature that GPU will probably not be able to support).

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/rowBasedExpressions.scala [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+          val cpuRunningTime = System.nanoTime - prepareArgsEnd
+          // Use log to record the eclipsed time for the Expression running before
+          // figuring out how to support Spark metrics in this expression.
+          logDebug(s"It took ${cpuRunningTime} ns to run Expression ${prettyName}, " +
+            s"and ${prepareArgsEnd - cpuExprStart} ns to get the input from children.")
+          GpuColumnVector.from(resultCol, dataType)
+        }
+      }
+    } catch {
+      case e: Exception =>
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/rowBasedExpressions.scala:105]: Why catch to just rethrow?

**Review by revans2 — COMMENTED** [review]
Generally it looks good., Do you have some performance numbers that we can look at?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/rowBasedExpressions.scala (hunk 1 above)
Comment (NVnavkumar) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/rowBasedExpressions.scala:105]: This was supposed to wrap the underlying exception in SparkException, but was not fully implemented, adding that now.

**Comment by NVnavkumar** [comment]
build

**Comment by NVnavkumar** [comment]
build

**Comment by NVnavkumar** [comment]
build

**Review by revans2 — DISMISSED** [review]

**Comment by NVnavkumar** [comment]
After more performance investigation and fixes, I believe this approach does not scale well. I have noticed the speedup consistently drops as the number of rows increases (this was after some performance optimizations and fixes):

image

image

These 2 tests involve several joins and only involve a single project or filter which would fallback to the GPU. I think this is happening because when the code is run on the CPU in the row-fallback approach, the operation is applied in serial on a GPU-sized batch designed to run in parallel. Therefore, I think another approach will be required to address this issue.

**Pull request closed without merging**